File: yasmin/src/yasmin/join_state.cpp
// Copyright (C) 2026 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "yasmin/join_state.hpp"

namespace yasmin {

JoinState::JoinState() : State({"joined"}), sync_id_(""), outcome_("joined") {}

JoinState::JoinState(const std::string &sync_id, const std::string &outcome)
    : State({outcome}), sync_id_(sync_id), outcome_(outcome) {}

void JoinState::configure() {
  if (this->has_parameter("sync_id")) {
    this->sync_id_ = this->get_parameter<std::string>("sync_id");
  }
  if (this->has_parameter("outcome")) {
    this->outcome_ = this->get_parameter<std::string>("outcome");
  }
}

void JoinState::set_barrier(RegionBarrier::SharedPtr barrier) {
  this->barrier_ = std::move(barrier);
}

const std::string &JoinState::get_sync_id() const { return this->sync_id_; }

RegionBarrier::SharedPtr JoinState::get_barrier() const {
  return this->barrier_;
}

std::string JoinState::execute(Blackboard::SharedPtr blackboard) {
  (void)blackboard;
  if (this->barrier_) {
    this->barrier_->arrive_and_wait();
  }
  return this->outcome_;
}

std::string JoinState::to_string() const {
  return std::string("JoinState [") + this->sync_id_ + "]";
}

} // namespace yasmin

File: yasmin_editor/yasmin_editor/model/join_state.py
# Copyright (C) 2026 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from yasmin_editor.dataclass_compat import dataclass

from .outcome import Outcome
from .state import State


@dataclass(slots=True, repr=False)
class JoinState(State):
    """Represents a YASMIN join-state leaf that synchronizes region threads."""

    sync_id: str = ""
    join_outcome: str = "joined"

    def __post_init__(self) -> None:
        if not self.outcomes:
            self.outcomes = [Outcome(name=self.join_outcome)]

    def __str__(self) -> str:
        parts = [f"JoinState({self.name})"]
        if self.sync_id:
            parts.append(f"sync_id={self.sync_id}")
        parts.append(f"outcome={self.join_outcome}")
        return " ".join(parts)

    __repr__ = __str__
